Fix invalid certificate page for a11y and translation.

adeelkhan opened the pull request:
Invalid certificate page has hardcoded lang='en' for
html, that would cause a11y issue for other languages user.
Some of screen reader text is not under translation. This patch
would fix both issues.

LEARNER-3110

adeelkhan commented:
@amangano-edx I guess removing lang attribute would fix a11y issue. It appears lang attribute is most used by search engines and screen readers. I guess its fine if the page is not crawled by search engines, but i dont know what impact removing this would create for screen readers. I can't find a way to test invalid certificate page for multiple language.

amangano-edx commented on lms/templates/certificates/accomplishment-base.html:
```diff
@@ -9,7 +9,7 @@
 %>
 
 <!DOCTYPE html>
-<html class="no-js" lang="en">
```
Removing this does not fix the issue. You need to set it dynamically to whatever the current language is, as otherwise screenreaders will not be able to properly read the translated strings.

amangano-edx replied:
Sanity Check: Is the LANGUAGE_CODE variable here dynamically set according to the current active language for the page, or is it the application default language?

adeelkhan replied:
@amangano-edx I can't able to find the place from where and how this LANGUAGE_CODE variable is injected into template. Cant able to find any such variable in context. I got idea and help from https://github.com/edx/edx-platform/blob/master/lms/templates/main.html#L27. Manually testing it shows the variable is correctly set for any active language when i change the user language via  http://localhost:18000/account/settings 
![lang_en](https://user-images.githubusercontent.com/373677/34724858-1b29741a-f571-11e7-80e5-4fa2760d13b8.png)
![lang_french](https://user-images.githubusercontent.com/373677/34724879-2233d2f0-f571-11e7-8ff6-5826d343f268.png)
![lang_arabic](https://user-images.githubusercontent.com/373677/34724880-2568f9dc-f571-11e7-8210-26e6e88f7567.png)

amangano-edx replied:
It might be safer to use `get_language` (https://docs.djangoproject.com/en/2.0/ref/utils/#django.utils.translation.get_language), and only fallback to LANGUAGE_CODE when `get_language` returns `None`. The Django docs make it clear what `get_language` does, so I think it'd be be better to rely on that than on the observed behavior of `LANGUAGE_CODE`.

amangano-edx replied:
Sorry, I think I misread your original comment. I think that if the main lms template uses LANGUAGE_CODE, then it is probably a good idea to use that here as well. Sorry for the confusion.

adeelkhan commented:
@amangano-edx Made the suggested changes please review.

amangano-edx commented on lms/templates/certificates/_accomplishment-rendering.html:
```diff
@@ -1,3 +1,4 @@
+<%page expression_filter="h"/>
```
Some of the strings in this file use the `{| h}` filter. Will that conflict with this (and result in double escaping)?

amangano-edx commented on lms/templates/certificates/_accomplishment-footer.html:
```diff
@@ -1,3 +1,4 @@
+<%page expression_filter="h"/>
```
Will this cause any issues for URLs used on this page? (Will urls containing & be escaped to &amp;)?

amangano-edx commented on lms/templates/certificates/invalid.html:
```diff
@@ -1,3 +1,4 @@
+<%page expression_filter="h"/>
 <%inherit file="accomplishment-base.html" />
 <%! from django.utils.translation import ugettext as _ %>
```
Line 15 in this file contains an aria-label attribute that should probably be translated.

amangano-edx commented:
This is looking better. Please make sure the verify that introducing the HTML filtering on each of those templates doesn't introduce any issues with double escaping or escaping of things like URLs/Javascript.

adeelkhan commented:
jenkins run quality

adeelkhan commented:
@amangano-edx I have verified and looked into any possibility of double escaping. Added javascript escaping where needed and removed html escaping for urls.  Used xss-linter to fix any javascript/html related escaping issue. I guess now it looks fine, but if you find anything missing or think might cause any side effects do mention thanks!

amangano-edx commented on lms/templates/certificates/_accomplishment-banner.html:
```diff
@@ -40,18 +41,18 @@
                     % if facebook_share_enabled:
                       <button class="action action-share-facebook btn-inverse btn-small icon-only" id="action-share-facebook"
                          onclick="FaceBook.share({
-                          share_text: '${facebook_share_text | escapejs}',
```
what's the difference between escapedjs and js_escaped_string?

adeelkhan replied:
Apparently looking into js_escaped_string code shows no difference. It working like its a wrapper to escapedjs. I had used in a previous work so thought about changing it too. As edx documentation mostly refers it http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/conventions/preventing_xss.html#javascript-context-in-mako

amangano-edx commented on lms/templates/certificates/_accomplishment-rendering.html:
```diff
@@ -46,7 +47,7 @@ <h3 class="accomplishment-signatories-title sr-only">${_("Noted by")}</h3>
                                 % for signatory in certificate_data.get('signatories', []):
                                 <div class="signatory">
-                                    <img class="signatory-signature" src="${static.url(signatory['signature_image_path'])}" alt="${signatory['name']}">
+                                    <img class="signatory-signature" src="${static.url(signatory['signature_image_path']) | n, decode.utf8}" alt="${signatory['name']} | n, decode.utf8">
```
I think the last call to `| n, decode.utf8` may have been placed there in error? It looks like it's part of the HTML and not the mako code.

adeelkhan replied:
Yep its copy/paste error. I'll fix that.

amangano-edx replied:
We still need to apply HTML escaping to the img source URL. Please remove the n filter. See http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/conventions/preventing_xss.html#url-context-and-escaping for more details.

amangano-edx commented on lms/templates/certificates/accomplishment-base.html:
```diff
@@ -1,15 +1,17 @@
 from django.utils.translation import get_language_bidi
 dir_rtl = 'rtl' if get_language_bidi() else 'ltr'
 course_mode_class = course_mode if course_mode else ''
+user_lang = get_language() if get_language() else LANGUAGE_CODE
```
To avoid calling `get_language()` twice, I think you can use `or`.
```
user_lang = get_language() or LANGUAGE_CODE
```

amangano-edx commented:
According to http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/conventions/preventing_xss.html#url-context-and-escaping, we still need to apply HTML-escaping to URLs, so I think we can remove the calls to the n filter that you added for the URLS.

amangano-edx commented on lms/templates/certificates/_about-edx.html:
```diff
@@ -1,10 +1,11 @@
         <p>${company_about_description}</p>
 
-        <a class="action" href="${company_about_url}">${company_about_urltext}</a>
+        <a class="action" href="${company_about_url | n, decode.utf8}">${company_about_urltext}</a>
```
According to http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/conventions/preventing_xss.html#url-context-and-escaping, we still need to HTML/Javascript-escape URLs that are added to the page. Therefore, we can remove the n filter here.

amangano-edx commented on lms/templates/certificates/_accomplishment-banner.html:
```diff
@@ -16,15 +17,15 @@
               };
               Logger.log('edx.certificate.shared', data);
-              window.open('${linked_in_url}');
+              window.open('${linked_in_url | n, decode.utf8}');
```
Please apply the js_escaped_string filter to this, as we still need to Javascript/HTML escape URLS. See 
http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/conventions/preventing_xss.html#url-context-and-escaping for more details.

adeelkhan replied:
Initially i did applied it, but it was breaking one of test. Inspecting into the issue and looking into the url shows that js escaping was creating the issue

without escaped url
---------------------------
``` http://www.linkedin.com/profile/add?_ed=0_0dPSPyS070e0HsE9HNz_13_d11_&amp;pfCertificationName=%C3%A9dX+Honor+Code+Certificate+for+refundable+course&amp;pfCertificationUrl=http%3A%2F%2Ftestserver%2Fcertificates%2Ff61a7536c1f64e93a468ced951805041&amp;source=o&amp;trk=unittest-testorg%2Frun1%2Frefundable_course_honor-dashboard ```

default html escape via <%page expression_filter="h"/>
-------------------------------------------------------------------------
``` http://www.linkedin.com/profile/add?_ed=0_0dPSPyS070e0HsE9HNz_13_d11_&pfCertificationName=%C3%A9dX+Honor+Code+Certificate+for+refundable+course&pfCertificationUrl=http%3A%2F%2Ftestserver%2Fcertificates%2Ff61a7536c1f64e93a468ced951805041 ```

javascript escaped via ${ some_var | n, js_escaped_string }
-------------------------------------------------------------------------------
``` http://www.linkedin.com/profile/add?_ed\\u003Dtest\\u0026pfCertificationName\\u003DTest+Site+Honor+Code+Certificate+for+refundable+course\\u0026pfCertificationUrl\\u003Dhttp%3A%2F%2Ftest\\u002Dsite.testserver%2Fcertificates%2F3cc0343f08734bbc918c8dff2f5487d4\\u0026source\\u003Do\\u0026trk\\u003Dunittest\\u002Dtestorg%2Frun1%2Frefundable_course_honor\\u002Ddashboard\ ```

adeelkhan replied:
@amangano-edx Removed js_escaped_string as it creates wrong linkedin url, which would not work. Now only html escaping.
```
>>> from django.utils.html import escapejs
>>> escapejs('http://something.com/?a=123&b=adeel+khan')
u'http://something.com/?a\\u003D123\\u0026b\\u003Dadeel+khan'
```

amangano-edx commented on lms/templates/certificates/_accomplishment-banner.html:
```diff
@@ -40,18 +41,18 @@
-                          description: '${_('Click the link to see my certificate.') | escapejs}'
+                          share_text: '${facebook_share_text | n, js_escaped_string}',
+                          share_link: '${share_url | n, decode.utf8}',
+                          picture_link: '${full_course_image_url | n, decode.utf8}',
```
Please apply the js_escaped_string filters to these URLs.

adeelkhan replied:
@amangano-edx js escaping would generate an incorrect url. Now only html escaping.

amangano-edx commented on lms/templates/certificates/_accomplishment-footer.html:
```diff
@@ -1,16 +1,17 @@
     <footer class="footer-app" role="contentinfo" id="company-info">
         <div class="footer-app-copyright">
-            <p class="copy copy-micro">${copyright_text}</p>
+            <p class="copy copy-micro">${copyright_text | n, decode.utf8 }</p>
```
Why is this exempt from the HTML filter?

adeelkhan replied:
Because the copyright was not properly appearing. Html character entities was getting escaped so had to remove escaping.
```
context['copyright_text'] = u'&copy; {year} {platform_name}. {reserved}.'.format(
        year=datetime.now(pytz.timezone(settings.TIME_ZONE)).year,
        platform_name=platform_name,
        reserved=reserved
    )
```

amangano-edx replied:
Does the XSS linter flag this? You might need to explicitly disable it for this line.

amangano-edx commented on lms/templates/certificates/_accomplishment-footer.html:
```diff
@@ -1,16 +1,17 @@
                 </li>
                 <li class="nav-item">
-                    <a class="action btn btn-small btn-link" href="${company_privacy_url}">${company_privacy_urltext}</a>
+                    <a class="action btn btn-small btn-link" href="${company_privacy_url | n, decode.utf8 }">${company_privacy_urltext}</a>
```
We still need to apply HTML escaping to these URLs. Please remove the n filter. See http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/conventions/preventing_xss.html#url-context-and-escaping for more details.

adeelkhan replied:
I'll check this document to see how i can add html escaping to all url.

amangano-edx replied:
If you remove the n filter, the default HTML escaping logic should be applied.

amangano-edx commented on lms/templates/certificates/_accomplishment-header.html:
```diff
@@ -1,11 +1,12 @@
-            <a class="logo" href="${logo_url}">
-                <img class="logo-img" src="${logo_src}" alt="${_('{platform_name} Home').format(platform_name=platform_name)}" />
+            <a class="logo" href="${logo_url | n, decode.utf8 }">
+                <img class="logo-img" src="${logo_src | n, decode.utf8 }" alt="${_('{platform_name} Home').format(platform_name=platform_name)}" />
```
We still need to apply HTML escaping to these URLs. Please remove the n filter. See http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/conventions/preventing_xss.html#url-context-and-escaping for more details.

amangano-edx commented on lms/templates/certificates/_accomplishment-rendering.html:
```diff
@@ -68,14 +69,14 @@ <h3 class="accomplishment-orgs-title sr-only">${_("Supported by the following or
                     <li class="wrapper-organization">
                         <div class="organization organization-edX">
-                            <img class="organization-logo" src="${logo_src}" alt="${platform_name}">
+                            <img class="organization-logo" src="${logo_src | n, decode.utf8}" alt="${platform_name}">
```
We still need to apply HTML escaping to the img source URL. Please remove the n filter. See http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/conventions/preventing_xss.html#url-context-and-escaping for more details.

amangano-edx commented on lms/templates/certificates/_accomplishment-rendering.html:
```diff
@@ -68,14 +69,14 @@ <h3 class="accomplishment-orgs-title sr-only">${_("Supported by the following or
                     <li class="wrapper-organization">
                         <div class="organization">
-                            <img class="organization-logo" src="${organization_logo.url}" alt="${platform_name}">
+                            <img class="organization-logo" src="${organization_logo.url | n, decode.utf8}" alt="${platform_name}">
```
We still need to apply HTML escaping to the img source URL. Please remove the n filter. See http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/conventions/preventing_xss.html#url-context-and-escaping for more details.

adeelkhan commented:
@amangano-edx Fixed rest of the things that you asked. Please review again. thanks

amangano-edx commented on lms/djangoapps/certificates/tests/test_webview_views.py:
```diff
@@ -273,8 +274,8 @@ def test_linkedin_share_url(self):
-            self.linkedin_url.format(params=urlencode(params)),
-            response.content
+            escape(self.linkedin_url.format(params=urlencode(params))),
+            unicode(response.content, 'utf-8')
```
Why do we need to explicitly pass 'utf-8' to `unicode`?

adeelkhan replied:
@amangano-edx  This test was throwing exception 
```
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 267: ordinal not in range(128)
```
The issue is js_escape filter is returning unicode string ,causing response.content byte string to be implicitly decoded to unicode string. As the response.content is having some unicode character the default codec 'ascii' is not able to decode it successfully. We have to explicitly decode via utf-8 before using assertIn. This is not needed in below test as it is not inserting any unicode. Simple ascii would get easily decoded implicitly via default codec.

Now the question is why the test was previously passing. My understanding is as there was no escaping, python was not forced to implicitly decode byte string.

amangano-edx commented on lms/djangoapps/certificates/tests/test_webview_views.py:
```diff
@@ -296,7 +297,7 @@ def test_linkedin_share_microsites(self):
         self.assertIn(
-            self.linkedin_url.format(params=urlencode(params)),
+            escape(self.linkedin_url.format(params=urlencode(params))),
             response.content
```
Why do we need to cast this to unicode on line 278, but not here?

amangano-edx commented on lms/djangoapps/certificates/tests/test_webview_views.py:
```diff
@@ -655,6 +656,23 @@ def test_html_view_for_invalid_certificate(self):
         self.assertIn("We cannot find a certificate with this URL or ID number.", response.content)
 
+    @override_settings(FEATURES=FEATURES_WITH_CERTS_ENABLED)
+    @override_settings(LANGUAGE_CODE='fr')
```
I'm not sure if this is testing that the lang attribute is set in a way that is truly dynamic, or if it's testing that the lang attribute is set to match the `LANGUAGE_CODE` setting.

amangano-edx commented on lms/templates/certificates/_accomplishment-banner.html:
```diff
@@ -40,18 +41,18 @@
-                          share_text: '${facebook_share_text | escapejs}',
+                          share_text: '${facebook_share_text | n, js_escaped_string}',
                           share_link: '${share_url}',
                           picture_link: '${full_course_image_url}',
```
We should apply the `js_escaped_string` filter to these URLs as well. I don't believe there should be any problem with javascript-escaping the URLs

amangano-edx commented on lms/templates/certificates/_accomplishment-rendering.html:
```diff
@@ -24,12 +25,12 @@
                         <span class="accomplishment-course hd-1 emphasized">
                             <span class="accomplishment-course-org">${accomplishment_copy_course_org}  </span>
-                            <span class="accomplishment-course-number">${course_number}</span>:
+                            <span class="accomplishment-course-number">${course_number | n, decode.utf8 }</span>:
```
Why is this exempt from the filter?

adeelkhan commented:
@amangano-edx Please when possible do review. Have made changes.

amangano-edx commented on lms/djangoapps/certificates/tests/test_webview_views.py:
```diff
@@ -655,6 +657,32 @@ def test_html_view_for_invalid_certificate(self):
+        self.cert.invalidate()
+
+        user_language = 'fr'
+        translation.activate(user_language)
```
Use `with translation.override` to ensure that the language is properly switched back between assertions and after the test finishes. See https://docs.djangoproject.com/en/2.0/ref/utils/#django.utils.translation.override

amangano-edx commented on lms/djangoapps/certificates/tests/test_webview_views.py:
```diff
@@ -655,6 +657,32 @@ def test_html_view_for_invalid_certificate(self):
+
+        user_language = 'fr'
+        translation.activate(user_language)
+        self.client.cookies[settings.LANGUAGE_COOKIE] = user_language
```
Why do you need to set the cookie if you've already changed the Django language to `user_language` using `translation.activate`/`translation.override`?

adeelkhan replied:
@amangano-edx translation.override is not working correctly for above test case. The language does get activated but the change doesn't last for client.get() call. I initially tried below snippet but no success
```
 with translation.override('fr'):
            self.assertEqual(get_language(), 'fr')
            response = self.client.get(test_url)
            self.assertIn('<html class="no-js" lang="fr">', response.content)
``` 
The first assert passes, but inspecting response.content shows the default lang.  translation.activate('user_language') only loads, but does't activates/switch in my case, so i need to either set the cookie or accept-header params to activate this.  Inspecting client.cookies dictionary in test showed "openedx-language-preference" key was empty. Setting this value causes django to activate language based on its value. I guess we are requesting a view so django evaluates language based on these rules. https://docs.djangoproject.com/en/2.0/topics/i18n/translation/#how-django-discovers-language-preference. Have removed translation.activation as it was not working at all.

amangano-edx commented on lms/djangoapps/certificates/views/webview.py:
```diff
@@ -173,7 +173,7 @@ def _update_context_with_basic_info(context, course_id, platform_name, configura
     context['document_title'] = _("Invalid Certificate")
 
     # Translators: The &amp; characters represent an ampersand character and can be ignored
```
This comment is no longer necessary.

amangano-edx commented on lms/templates/certificates/_accomplishment-banner.html:
```diff
@@ -40,18 +41,18 @@
+                          picture_link: '${full_course_image_url | n, js_escaped_string}',
+                          description: '${_('Click the link to see my certificate.') | n, js_escaped_string}'
                           });">
                           <span class="icon fa fa-facebook-official" aria-hidden="true"></span>
```
There's some twitter button JS below this on lines 62-68 that you should also apply the `js_escaped_string` filter to.

adeelkhan replied:
@amangano-edx I guess you mean only to JS code as the rest would be html escaped.

amangano-edx commented on lms/templates/certificates/_accomplishment-rendering.html:
```diff
@@ -24,12 +25,12 @@
                         <span class="accomplishment-course hd-1 emphasized">
                             <span class="accomplishment-course-org">${accomplishment_copy_course_org}  </span>
-                            <span class="accomplishment-course-number">${course_number}</span>:
+                            <span class="accomplishment-course-number">${course_number }</span>:
```
Nit: There's extra whitespace after `course_number`.

amangano-edx commented on lms/templates/certificates/accomplishment-base.html:
```diff
@@ -1,15 +1,17 @@
+<%page expression_filter="h"/>
 <%namespace name='static' file='/static_content.html'/>
-<%! from django.utils.translation import ugettext as _ %>
+<%! from django.utils.translation import ugettext as _ , get_language %>
```
Nit: there's whitespace between `_` and the comma.

amangano-edx commented on lms/templates/certificates/server-error.html:
```diff
@@ -1,3 +1,4 @@
+<%page expression_filter="h"/>
 <%! from django.utils.translation import ugettext as _ %>
 <%inherit file="../main.html" />
 
```
The translation on line 17 contains an html entity that will be double escaped.

amangano-edx commented:
I found a few more issues that should be addressed before this is merged. This is very close to being ready.

adeelkhan commented:
jenkins run all

amangano-edx commented on lms/djangoapps/certificates/tests/test_webview_views.py:
```diff
@@ -655,6 +657,30 @@ def test_html_view_for_invalid_certificate(self):
+        )
+
+        # invalidate certificate and verify that "Cannot Find Certificate" is returned
+        self.cert.invalidate()
```
Is this necessary for this test?

amangano-edx commented on lms/djangoapps/certificates/tests/test_webview_views.py:
```diff
@@ -655,6 +657,30 @@ def test_html_view_for_invalid_certificate(self):
+            course_id=unicode(self.course.id)
+        )
+
+        # invalidate certificate and verify that "Cannot Find Certificate" is returned
```
It doesn't look like we're actually checking for this string anywhere.

amangano-edx commented on lms/djangoapps/certificates/tests/test_webview_views.py:
```diff
@@ -655,6 +657,30 @@ def test_html_view_for_invalid_certificate(self):
+        self.cert.invalidate()
+
+        user_language = 'fr'
+        self.client.cookies[settings.LANGUAGE_COOKIE] = user_language
```
I wonder if `translation.override` wasn't working because we hadn't created a `DarkLangConfig` for this test?
 I'm curious how setting the cookie works without that config, as the DarkLang middleware should only allow released languages to be resolved from the cookie. See https://github.com/edx/edx-platform/blob/master/lms/djangoapps/certificates/tests/test_webview_views.py#L1118 for an example DarkLangConfig.

amangano-edx replied:
I spoke with Cale about DarkLang, and learned that it's not necessary for these tests. Please disregard my comment. Your test does not need to change.

adeelkhan commented:
@amangano-edx Made the changes. Please review. thanks

amangano-edx commented on lms/templates/certificates/_accomplishment-banner.html:
```diff
@@ -61,7 +62,8 @@ <h3 class="sr-only">${_("Print or share your certificate:")}</h3>
                         class="action action-share-twitter btn-inverse btn-small icon-only"
                         title="${_('Share on Twitter')}"
-                        onclick="popupWindow('${twitter_url}', 'tweetWindow', 640, 480); return false;">
+                        ## xss-lint: disable=mako-invalid-html-filter
```
Please include the comment explaining why the linter is being disabled, as suggested by Robert.
`## xss-lint: disable=mako-invalid-html-filter  Linter mistakes onclick as html.`
